Update ProcessorTest

jomarko opened the pull request:
This is a follow up PR:
- for https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/229/files
- and https://issues.redhat.com/browse/MGDOBR-797

kelvah commented:
@jomarko FYI there's [also this ticket](https://issues.redhat.com/browse/MGDOBR-1055) covering some of the processors tests

jomarko commented:
@kelvah @ppadti ok, I am not going to add new tests, just fix existing

jomarko commented:
I will rebase this PR once the original #229 PR is merged

jomarko commented:
Decided to rebase PR right now. The changes are not dependent on #229

jomarko commented:
Unrelated `InstanceCreation` test failed. It is being solved by #226 and #230

```
<testsuites name="Mocha Tests" time="145.3570" tests="4" failures="1">
  <testsuite name="Root Suite" timestamp="2022-10-20T05:34:43" tests="0" file="integration/instance/InstanceCreation.ts" time="0.0000" failures="0">
  </testsuite>
  <testsuite name="the &apos;Create a SE instance&apos; Modal" timestamp="2022-10-20T05:34:43" tests="4" time="145.3570" failures="1">
    <testcase name="the &apos;Create a SE instance&apos; Modal Submit" time="0.0000" classname="Submit">
      <failure message="Timed out retrying after 120000ms: `cy.should()` failed because this element is detached from the DOM.

`&lt;div data-ouia-component-id=&quot;steps-count&quot; data-ouia-component-type=&quot;QE/StackItem&quot; class=&quot;pf-l-stack__item&quot;&gt;1 of 3 ...&lt;/div&gt;`

Cypress requires elements be attached in the DOM to interact with them.

The previous command that ran was:

  &gt; `cy.ouiaId()`

This DOM element likely became detached somewhere between the previous and current command.

Common situations why this happens:
  - Your JS framework re-rendered asynchronously
  - Your app code reacted to an event firing and removed the element

You typically need to re-query for the element or add &apos;guards&apos; which delay Cypress from running new commands.

https://on.cypress.io/element-has-detached-from-dom" type="CypressError"><![CDATA[CypressError: Timed out retrying after 120000ms: `cy.should()` failed because this element is detached from the DOM.

`<div data-ouia-component-id="steps-count" data-ouia-component-type="QE/StackItem" class="pf-l-stack__item">1 of 3 ...</div>`

Cypress requires elements be attached in the DOM to interact with them.

The previous command that ran was:

  > `cy.ouiaId()`

This DOM element likely became detached somewhere between the previous and current command.

Common situations why this happens:
  - Your JS framework re-rendered asynchronously
  - Your app code reacted to an event firing and removed the element

You typically need to re-query for the element or add 'guards' which delay Cypress from running new commands.
```

vpellegrino commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -18,7 +15,14 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
       it("Sink processor", () => {
         const processorName: string = "Sink processor";
-        const action = ["Slack", "dev-action", "https://test.app.com/item"];
+        const action = {
```
what about having constants used as key in this `action` object?
example:
```
var CONSTANT_A = 0, CONSTANT_B = 1
var state_machine = {
    [CONSTANT_A]: function () {
        return 'a'
    },
    [CONSTANT_B]: function () {
        return 'b'
    }
};
```

in this way you can use them also in lines 72, 84, 91

vpellegrino replied:
Square brackets are used inside the object definition to have variables/context inside the object's key (https://stackoverflow.com/a/68796045).
Once you have defined a constant for each string, you can use the constant directly in the mentioned rows.

sjamboro replied:
@jomarko in case that you want to use some constants for Action/Source/Filters labels can you think about importing them from Utils? My future thoughts: it would be nice to import and read these objects: https://github.com/jomarko/sandbox-ui/blob/PR-MGDOBR-1148/mocked-api/schemasData.ts#L240-L290

sjamboro reviewed (commented):
IIUC this PR is about removing a flag which causes that several tests were skipped. I can see some refactoring in the action/source data format. It seems that execution of these tests are not change. 

@jomarko thank you for refactoring of this test suite. It is more readable. I have left some comments.

vpellegrino reviewed (approved):
Just left one comment about extracting to constants few strings, apart from that it looks good to me

sjamboro commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -18,7 +15,14 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
-        const action = ["Slack", "dev-action", "https://test.app.com/item"];
+        const action = {
+          "Action type": "Slack",
+          Channel: "dev-action",
```
To be honest ... in future I would like to see here some function which will read information directly from action definition.

jomarko replied:
@sjamboro Thank you for this observation. It is related to prettier. https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/232/commits/32886921119f58e878d3907a25a96e2398652cfc

jomarko commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -406,14 +409,19 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
 
       cy.ouiaId("action-section", "PF4/Text").should("have.text", "Action");
       cy.get("[class='pf-c-description-list__term']")
```
I sent unification in a separate commit. Lets see CI

jomarko commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -18,7 +15,14 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
+          "Icon URL": "Property not configured",
+          Username: "Property not configured",
+        };
         const filters = [
```
looking into it

sjamboro commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -462,22 +473,28 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
+      filters: { [key in "Key" | "Type" | "Value"]: string }[]
+    ) {
       cy.ouiaId("filters-section", "PF4/Text").should("have.text", "Filters");
-      if (filters?.length > 0 && filters[0].length > 0) {
```
Is there any reason why you have removed this condition? Now we are sure that filters has only three types of keys (Key, Type, Value) but someone could input {Key: "data.name", Value: "John"} in such case the code will throw an expection. Is it the expected behaviour? I understand that the previous structure of code is not ideal.

jomarko replied:
I agree this is a place for future improvement. However I do not want to do it as part of this PR, that is already quite large. If I understand correctly the code, even before my refactoring, we didn't validated filter 'deeply'. To me it seems that `input {Key: "data.name", Value: "John"}` was 'valid'  also before my refactoring.

sjamboro commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -310,43 +327,38 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
-            return;
-          }
+      it("Add filter row", () => {
+        filters[0].Key = "data.surname";
```
@jomarko There could be a problem

sjamboro replied:
Do you think that there could be: `filters[1]["Key"]`?

jomarko replied:
I am not `TS` expert programmer, could you please clarify your concern?

is the problem index I used `0` vs `1`  , or `.Key` vs `["Key"]`?

sjamboro replied:
@jomarko  I mean that `.Key` might caused that problem. I am not TS expert too but based on my Java and JS experience:
1.`.Key` reference the property of an object which is saved in the `fitlers` array under index `0`
2. `["Key"]` reference the item in an array/map structure which key is `"Key"` this array/map is saved in the `filters` array under index '1'.

sjamboro replied:
@jomarko the error message is <failure message="Cannot set properties of undefined (setting 'Key')" type="TypeError">

jomarko replied:
ok, pushing separate commit with `["xxx"]` syntax

sjamboro commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -225,50 +240,50 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
-        it("Add filter row", () => {
-          filters[1] = ["data.surname", "String equals", "White"];
+      it("Add filter row", () => {
+        filters[1].Key = "data.surname";
```
@jomarko there could be a problem

sjamboro replied:
Do you think that there could be: `filters[1]["Key"]`?

jomarko replied:
similar question as above

sjamboro replied:
@jomarko  I mean that `.Key` might caused that problem. I am not TS expert too but based on my Java and JS experience:
1.`.Key` reference the property of an object which is saved in the `fitlers` array under index `1`
2. `["Key"]` reference the item in an array/map structure which key is `"Key"` this array/map is saved in the `filters` array under index `1`.

sjamboro replied:
@jomarko It seems that problems is still here. Do you think that this could help: `filters.push({ Key: "data.name", Type: "String equals", Value: "John" });`?

sjamboro replied:
> @jomarko It seems that problems is still here. Do you think that this could help: `filters.push({ Key: "data.name", Type: "String equals", Value: "John" });`?

The failure was moved. The assertation of the new filter row failed >>> "undefined".

jomarko replied:
@sjamboro done in a separate commit

sjamboro reviewed (commented):
LGTM ... thank you that you have enabled "add filter row" tests.

Unfortunately, checks are failing so can you take a look at it?
![image](https://user-images.githubusercontent.com/1536609/197153568-3f2f69a0-4ca8-41a1-8ec1-bc04a688074f.png)
```
<testsuite name="Edit Sink Processors" timestamp="2022-10-20T15:20:25" tests="2" time="6.2780" failures="1">
<testcase name="Processor Test Edit Sink Processors Assert and cancel edit form" time="6.2780" classname="Assert and cancel edit form"> </testcase>
<testcase name="Processor Test Edit Sink Processors Add filter row" time="0.0000" classname="Add filter row">
<failure message="Cannot set properties of undefined (setting 'Key')" type="TypeError">
<![CDATA[ TypeError: Cannot set properties of undefined (setting 'Key') at Context.eval (webpack:///./integration/ProcessorTest.ts:244:22) ]]>
</failure>
</testcase>
</testsuite>
<testsuite name="Edit Source Processors" timestamp="2022-10-20T15:20:36" tests="2" time="7.1440" failures="1">
<testcase name="Processor Test Edit Source Processors Assert and cancel edit form" time="7.1440" classname="Assert and cancel edit form"> </testcase>
<testcase name="Processor Test Edit Source Processors Add filter row" time="0.0000" classname="Add filter row">
<failure message="Cannot set properties of undefined (setting 'Key')" type="TypeError">
<![CDATA[ TypeError: Cannot set properties of undefined (setting 'Key') at Context.eval (webpack:///./integration/ProcessorTest.ts:331:22) ]]>
</failure>
</testcase>
</testsuite>
```
It seems that filter structure does not work correctly. Can you take a look at it?

sjamboro commented on cypress/integration/ProcessorTest.ts:
```diff
@@ -18,10 +15,17 @@ onlyOn(isEnvironmentType(EnvType.Mocked), () => {
+          "Icon URL": "Property not configured",
+          "Username": "Property not configured",
+        };
+        const filters: { [key in "Key" | "Type" | "Value"]: string }[] = [
```
I think that this definiton could not be interpated as an property of an object (`.Key` notation)

jomarko replied:
not sure I understand, could you please try to clarify?

sjamboro replied:
@jomarko I have just highligths the definition of the `filters` variable which is mentioned in the previous notes. I am not sure but this definition says that you can use only `filtres["Key"][1]`, isn't it?

sjamboro reviewed (approved):
LGTM ... thank you ... this PR brings more readable code

kelvah commented on .prettierrc:
```diff
@@ -1,5 +1,6 @@
   "singleQuote": false,
-  "semi": true
+  "semi": true,
+  "quoteProps": "consistent"
```
I was OK with the default configuration, only requiring quotes when needed. But let's give this a try.

jomarko commented:
Waiting for https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/247
